util/netevent.c

```diff
@@ -755,6 +756,66 @@ static int udp_recv_needs_log(int err)
 	return 1;
 }
 
+/** Parses the PROXYv2 header from buf and updates the comm_reply struct.
+ *  Returns 1 on success, 0 on failure. */
+static int consume_pp2_header(struct sldns_buffer* buf, struct comm_reply* rep,
+	int stream) {
+	size_t size;
+	struct pp2_header *header = pp2_read_header(buf);
+	if(header == NULL) return 0;
+	size = PP2_HEADER_SIZE + ntohs(header->len);
+	if((header->ver_cmd & 0xF) == PP2_CMD_LOCAL) {
+		/* A connection from the proxy itself.
+		 * No need to do anything with addresses. */
+		goto done;
+	}
+	if(header->fam_prot == 0x00) {
+		/* Unspecified family and protocol. This could be used for
+		 * health checks by proxies.
+		 * No need to do anything with addresses. */
+		goto done;
+	}
+	/* Read the proxied address */
+	switch(header->fam_prot) {
+		case 0x11: /* AF_INET|STREAM */
+		case 0x12: /* AF_INET|DGRAM */
+			{
+			struct sockaddr_in* addr =
+				(struct sockaddr_in*)&rep->client_addr;
+			addr->sin_family = AF_INET;
+			addr->sin_addr.s_addr = header->addr.addr4.src_addr;
+			addr->sin_port = header->addr.addr4.src_port;
+			rep->client_addrlen = (socklen_t)sizeof(struct sockaddr_in);
+			}
+			/* Ignore the destination address; it should be us. */
+			break;
+		case 0x21: /* AF_INET6|STREAM */
+		case 0x22: /* AF_INET6|DGRAM */
+			{
+			struct sockaddr_in6* addr =
+				(struct sockaddr_in6*)&rep->client_addr;
+			addr->sin6_family = AF_INET6;
```

Comment:
The IPv6 sockaddr contains flowinfo and sometimes scope_id, and these should be set to a value. The change initialises it.

```suggestion
			struct sockaddr_in6* addr =
				(struct sockaddr_in6*)&rep->client_addr;
			memset(addr, 0, sizeof(*addr));
			addr->sin6_family = AF_INET6;
```